src/test/scala/chiseltest/coverage/SimulatorCoverageTest.scala

```diff
@@ -0,0 +1,85 @@
+import chisel3._
+
+/** Ensure that all simulators give the same coverage feedback when run with the same tests.
+  * To implement this for a particular simulator, just override the `backend` annotation.
+  * */
+abstract class SimulatorCoverageTest(name: String, backend: BackendAnnotation) extends AnyFlatSpec with ChiselScalatestTester {
+  behavior of s"$name Coverage Collection"
+  private def noAutoCov: AnnotationSeq = Seq(backend)
+
+  it should "report count for all user cover points (no submodules)" in {
+    val r = test(new Test1Module(withSubmodules = false)).withAnnotations(noAutoCov) { dut =>
+      dut.clock.step()
+    }
+    val cov = getCoverage(r)
+
+    // the cover point in the main module are not prefixed
+    assert(cov.keys.toList == List("cover_0"))
+
+    // since we executed one step (+ the implicit reset) and all inputs are zero by default, we expect the count to be 3
+    assert(cov("cover_0") == 3)
+  }
+
+  it should "report count for all user cover points (with submodules)" in {
+    val r = test(new Test1Module(withSubmodules = true)).withAnnotations(noAutoCov) { dut =>
+      dut.clock.step()
+    }
+    val cov = getCoverage(r)
+
+    // the cover point in the main module are not prefixed, but the one in the child module are
+    assert(cov.keys.toList.sorted == List("c0.cover_0", "c1.cover_0", "cover_0"))
+
+    // since we executed one step (+ the implicit reset) and all inputs are zero by default, we expect the count to be 3
+    assert(cov("cover_0") == 3)
+  }
+
+  it should "allow us to estimate pi" in {
+    val rand = new scala.util.Random(0)
+    val r = test(new PiEstimator).withAnnotations(noAutoCov) { dut =>
+      (0 until 1000).foreach { _ =>
+        dut.x.poke(BigInt(8, rand).S)
+        dut.y.poke(BigInt(8, rand).S)
+        dut.clock.step()
+      }
+    }
+
+    val cov = getCoverage(r)
+    val inCircle = cov("cover_0")
+    val inRectangle = cov("cover_1")
+    val pi = 4.0 * inCircle.toDouble / inRectangle.toDouble
+    val error = math.abs(pi - math.Pi)
+    assert(error < 0.0329)
+  }
+
+  private def getCoverage(annos: AnnotationSeq): Map[String, Long] = {
+    val coverage = annos.collect { case a: TestCoverage => a.counts }
+    assert(coverage.size == 1, "Expecting exactly one coverage map!")
+    coverage.head.toMap
+  }
+}
+
+class TreadleCoverageTest extends SimulatorCoverageTest("Treadle", TreadleBackendAnnotation) {}
+class VerilatorCoverageTest extends SimulatorCoverageTest("Verilator", VerilatorBackendAnnotation) {}
+
+private class PiEstimator extends Module {
+  val x = IO(Input(SInt(8.W)))
+  val y = IO(Input(SInt(8.W)))
+  val inCircle = IO(Output(Bool()))
+  val inRectangle = IO(Output(Bool()))
+  val radius = 100
+  inCircle := (x * x + y * y) <= (radius * radius).S
+  inRectangle := (x <= radius.S && x >= -radius.S) && (y <= radius.S && y >= -radius.S)
+  chisel3.experimental.verification.cover(inCircle)
+  chisel3.experimental.verification.cover(inRectangle)
+}
```

Comment:
missing newline

Reply:
I was confused for a bit why `scalafmt` didn't fix this. Turns out it does not run on the `test` directory, only on `main`.

Reply:
fixed